zerver/lib/markdown/__init__.py

```diff
@@ -1884,11 +1884,28 @@ def handleMatch(  # type: ignore[override] # supertype incompatible with supersu
 
             wildcard = mention.user_mention_matches_wildcard(name)
 
-            id_syntax_match = re.match(r"(.+)?\|(?P<user_id>\d+)$", name)
+            id_syntax_match = re.match(r"(?P<full_name>.+)?\|(?P<user_id>\d+)$", name)
             if id_syntax_match:
+                # Most of the logic supposed to be here is simplified
+                # because of the careful approach we take while
+                # initializing mention_data in get_possible_mentions_info.
+
+                # This block deals with @**|id** and @**name|id** syntax.
+                # In case of @**name|id**, we specifically check that name
+                # matches the full_name of user in mention data, this avoids cases
+                # like @**invalid_name|valid_id** where user with valid_id might
+                # be present in mention_data because of some other @**|valid_id**
+                # or @**valid_name|valid_id** mention in same message.
+
+                full_name = id_syntax_match.group("full_name")
                 id = int(id_syntax_match.group("user_id"))
                 user = db_data["mention_data"].get_user_by_id(id)
+                if full_name:
+                    # for **name|id** syntax
```

Comment:
This should say "For **name|id** syntax, we require that the name is correct".

Reply:
Made these changes!